NPE occurs when execute create database with postgres db

## Bug Report

### Which version of ShardingSphere did you use?

[e504221](https://github.com/apache/shardingsphere/commit/e504221d857b87744a040fda7cfb493d25a1e855)

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

ShardingSphere-Proxy

### Expected behavior

Execute following sql successfully.

```sql
postgres=> create database test1;
```

### Actual behavior

```
[ERROR] 2023-06-15 20:18:14.421 [pool-8-thread-1] c.g.common.eventbus.EventBus.default - Exception thrown by subscriber method renew(org.apache.shardingsphere.mode.manager.cluster.coordinator.registry.metadata.event.DatabaseAddedEvent) on subscriber org.apache.shardingsphere.mode.manager.cluster.coordinator.subscriber.ResourceMetaDataChangedSubscriber@dd20ebc when dispatching event: org.apache.shardingsphere.mode.manager.cluster.coordinator.registry.metadata.event.DatabaseAddedEvent@753adf4d
java.lang.NullPointerException: null
	at org.apache.shardingsphere.sqlfederation.optimizer.context.planner.OptimizerPlannerContextFactory.create(OptimizerPlannerContextFactory.java:79)
	at org.apache.shardingsphere.sqlfederation.rule.SQLFederationRule.alterDatabase(SQLFederationRule.java:50)
```

### Reason analyze (If you can)

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

### Example codes for reproduce this issue (such as a github link).

--- a/kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/context/OptimizerContext.java
+++ b/kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/context/OptimizerContext.java
@@ -49,6 +49,25 @@ public final class OptimizerContext {
     }
     
     /**
+     * Put parser context.
+     *
+     * @param databaseName database name
+     * @param parserContext parser context
+     */
+    public void putParserContext(final String databaseName, final OptimizerParserContext parserContext) {
+        parserContexts.put(databaseName.toLowerCase(), parserContext);
+    }
+    
+    /**
+     * Remove parser context.
+     *
+     * @param databaseName database name
+     */
+    public void removeParserContext(final String databaseName) {
+        parserContexts.remove(databaseName.toLowerCase());
+    }
+    
+    /**
      * Get planner context.
      *
      * @param databaseName database name

--- a/kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/rule/SQLFederationRule.java
+++ b/kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/rule/SQLFederationRule.java
@@ -25,6 +25,8 @@ import org.apache.shardingsphere.infra.rule.identifier.type.MetaDataHeldRule;
 import org.apache.shardingsphere.sqlfederation.api.config.SQLFederationRuleConfiguration;
 import org.apache.shardingsphere.sqlfederation.compiler.context.OptimizerContext;
 import org.apache.shardingsphere.sqlfederation.compiler.context.OptimizerContextFactory;
+import org.apache.shardingsphere.sqlfederation.compiler.context.parser.OptimizerParserContext;
+import org.apache.shardingsphere.sqlfederation.compiler.context.parser.OptimizerParserContextFactory;
 import org.apache.shardingsphere.sqlfederation.compiler.context.planner.OptimizerPlannerContext;
 import org.apache.shardingsphere.sqlfederation.compiler.context.planner.OptimizerPlannerContextFactory;
 
@@ -47,12 +49,15 @@ public final class SQLFederationRule implements GlobalRule, MetaDataHeldRule {
     
     @Override
     public void alterDatabase(final ShardingSphereDatabase database) {
-        OptimizerPlannerContext plannerContext = OptimizerPlannerContextFactory.create(database, optimizerContext.getParserContext(database.getName()), optimizerContext.getSqlParserRule());
+        OptimizerParserContext parserContext = OptimizerParserContextFactory.create(database.getProtocolType());
+        optimizerContext.putParserContext(database.getName(), parserContext);
+        OptimizerPlannerContext plannerContext = OptimizerPlannerContextFactory.create(database, parserContext, optimizerContext.getSqlParserRule());
         optimizerContext.putPlannerContext(database.getName(), plannerContext);
     }
     
     @Override
     public void dropDatabase(final String databaseName) {
+        optimizerContext.removeParserContext(databaseName);
         optimizerContext.removePlannerContext(databaseName);
     }
     

--- a/test/e2e/sql/src/test/java/org/apache/shardingsphere/test/e2e/engine/type/RDLE2EIT.java
+++ b/test/e2e/sql/src/test/java/org/apache/shardingsphere/test/e2e/engine/type/RDLE2EIT.java
@@ -52,7 +52,7 @@ import static org.junit.jupiter.api.Assertions.assertTrue;
 
 @E2ETestCaseSettings(SQLCommandType.RDL)
 class RDLE2EIT {
-
+    
     private static final Map<String, Boolean> INITIALIZED = new ConcurrentHashMap<>();
     
     @ParameterizedTest(name = "{0}")

